mender-update/http_resumer/http_resumer.cpp

```diff
@@ -0,0 +1,482 @@
+		// the user will get different headers on header and body handlers, representing (somehow)
+		// what the resumer has been doing in its behalf.
+		auto resp = exp_resp.value();
+		resumer_client->response_->status_code_ = resp->GetStatusCode();
+		resumer_client->response_->status_message_ = resp->GetStatusMessage();
+		for (auto header : resp->GetHeaders()) {
+			resumer_client->response_->headers_[header.first] = header.second;
+		}
+
+		// Finished, call the user handler \o/
+		resumer_client->Disable();
+		resumer_client->CallUserHandler(resumer_client->response_);
+		resumer_client->logger_.Debug("Download resumed and completed successfully");
+	}
+}
+
+DownloadResumerAsyncReader::~DownloadResumerAsyncReader() {
+	Cancel();
+}
+
+void DownloadResumerAsyncReader::Cancel() {
+	if (!*cancelled_) {
+		inner_reader_->Cancel();
+	}
+}
+
+error::Error DownloadResumerAsyncReader::AsyncRead(
+	vector<uint8_t>::iterator start, vector<uint8_t>::iterator end, io::AsyncIoHandler handler) {
+	auto resumer_client = resumer_client_.lock();
+	if (!resumer_client || *cancelled_) {
+		return error::MakeError(
+			error::ProgrammingError,
+			"DownloadResumerAsyncReader::AsyncRead called after stream is destroyed");
+	}
+	// Save user parameters for further resumes of the body read
+	resumer_client->last_read_ = {.start = start, .end = end, .handler = handler};
+	return AsyncReadResume();
+}
+
+error::Error DownloadResumerAsyncReader::AsyncReadResume() {
+	auto resumer_client = resumer_client_.lock();
+	if (!resumer_client) {
+		return error::MakeError(
+			error::ProgrammingError,
+			"DownloadResumerAsyncReader::AsyncReadResume called after client is destroyed");
+	}
+	return inner_reader_->AsyncRead(
+		resumer_client->last_read_.start,
+		resumer_client->last_read_.end,
+		[this](io::ExpectedSize result) {
+			if (!result) {
+				inner_reader_.reset();
+
+				logger_.Warning(
+					"Reading error, a new request will be re-scheduled. "
+					+ result.error().String());
+			} else {
+				resumer_state_->offset += result.value();
+				logger_.Debug("read " + to_string(result.value()) + " bytes");
+				auto resumer_client = resumer_client_.lock();
+				if (resumer_client) {
+					resumer_client->last_read_.handler(result);
+				} else {
+					logger_.Error(
+						"AsyncRead finish handler called after resumer client has been destroyed.");
+				}
+			}
+		});
+}
+
+DownloadResumerClient::DownloadResumerClient(
+	const http::ClientConfig &config, events::EventLoop &event_loop) :
+	client_(config, event_loop, "http_resumer:client"),
+	logger_ {"http_resumer:client"},
+	cancelled_ {make_shared<bool>(true)},
+	retry_ {.backoff = http::ExponentialBackoff(chrono::seconds(1), 10), .wait_timer = event_loop} {
+	resumer_state_ = make_shared<DownloadResumerClientState>();
+}
+
+DownloadResumerClient::~DownloadResumerClient() {
+	if (!*cancelled_) {
+		logger_.Warning("DownloadResumerClient destroyed while request is still active!");
+	}
+	client_.Cancel();
```

Comment:
As mentioned in the reader, it might be dangerous to call this here without disabling handler calling first. Maybe calling `Disable()` is enough, and the rest can be handled by the destructors in each object?

Reply:
I reworked all of these (and removed `Disable` ambiguous method).